README.md
# dmon

dmon is a WAD analysis tool that estimates the toughness of DOOM maps. It does this by adding up all health bonuses, ammo pickups and monster counts, then it compares those numbers against a preset baseline. This is useful for balancing your maps to closely match the said baseline.

The default baseline is DOOM II, however this can changed with options. See USAGE below.

# INSTALLATION

dmon is written in Python and targets version `3.8`, however it does run under Python `2.7` if you absolutely need it to.


*GNU / Linux or Mac*

You will need [Python](https://www.python.org/) installed.

Download and extract the latest source from [dmon/releases](https://github.com/wesleywerner/dmon/releases/latest), then run `python dmon.py` to invoke the magic.

You can alias this command in your favorite shell to make this command available anywhere:

```
alias dmon='python ~/path_to/dmon.py'
```

*Windows*

You can run from source or use the pre-built executable:

*Windows Executable*

Download and extract the latest build from [dmon/releases](https://github.com/wesleywerner/dmon/releases/latest). Create a new application shortcut with the below command. It runs a console window with dmon in its path:

```
cmd.exe /k set PATH=C:\Path\To\dmon-1.1_win;%PATH%
```

*Windows Source*

To run from source you need [Python](https://www.python.org/). Either install the official package, or use [WinPython](http://winpython.github.io) for an install-free experience. *Be sure to include Python in your PATH environment*, or use the `WinPython Command Prompt`.

Add the extracted Dmon `src` directory to your `PATH` environment variable and run `dmon`. There is a `dmon.bat` file that acts as the launcher to run from source in Windows.

# ISSUES

Please report any issues in the [issue tracker](https://github.com/wesleywerner/dmon/issues) or on the Doomworld forums.

# USAGE

```
$ dmon -h
Dmon

Usage:
  dmon <wad> [options]
  dmon <wad> <pattern> [options]
  dmon (-h | --help | --about | --license)

Options:
  -h --help                 Show this screen.
  --version                 Show current version.
  --license                 Print the license.
  --about                   Show how this all works.
  <pattern>                 match map name (E1M2 or MAP02)
                            '?' and '*' are wildcards eg:
                                E1M* (all episode 1)
                                MAP0[135] (MAP01, MAP03 and MAP05)
  -a, --average             Combine results from multiple maps into
                            average amounts.
  -u, --bonus               Include berserk, soulsphere and megesphere
                            as data points.
  -f <fmt>, --format=<fmt>  Set output format to: csv, json, dump
  -b, --baseline=<bl>       Set comparison baseline [Default: DOOM2]
                            where <bl> can be:
                                DOOM, DOOM2, SIGIL,
                                AV (Alien Vendetta),
                                HR (HELL REVEALED),
                                DTWID (Doom the Way id Did)
  -c, --compare             Show actual/baseline values for comparison.
  -d, --diff                Display the difference between actual and
                            baseline values.
  -x, --fixed               Display fixed-point numbers (otherwise
                            numbers are rounded up).
  -l, --legend              Print the recommendation legend at the end.
```

# ABOUT

```
$ dmon --about

WAD analysis tool that estimates the toughness of DOOM maps

It queries one or more maps in a WAD/PWAD and prints out some crunched
numbers to try and answer those hard questions.

It tries to help map creators answer:
 * The percentage of monsters that use hitscan attacks
 * How much health and armor pickups there are, relative to how much
   damage monsters can do to the player
 * The amount of damage the player can inflict on monsters, given the 
   amount of ammo

The following data points are gathered:
 1) sum damage and hit points of monsters, and how many use hitscan attacks
 2) sum health points for health bonus(1), stimpacks(10), medkits(25)
 3) sum armor points for armor bonus(1), green armor(100), blue armor(200)
 4) sum of ammunition rounds (including weapons and enemy drops)

These values are then derived:
 * Health and Armor bonus over median monster attack damage
 * Ammunitions damage over monster hit points

A shell damage of 1 indicates enough shells to kill each monster at least once,
assuming the median amount of damage is inflicted.

Bullet damage is the median for the Chaingun, per tap of the trigger.
Shell damage is the median for the shotgun, per shot.
Rocket damage is the median for the rocket launcher, per rocket.
Plasma damage is the median for the plasma gun, per shot.

To help put these numbers in perspective, they are compared to a baseline and
dmon makes recommendations based on the outcome.
The default baseline is DOOM2 and can be changed with the --baseline option.
If a metric falls outside the baseline a recommendation is made.
You can print the legend of these recommendations with the --legend option.

Note: Berserk(100), Soul sphere(100) and Megasphere(200) items are
not counted by default. To count these use the "--bonus" option.

Dmon uses Omgifol to read WAD files (see --license).

Indemnity: The metrics used here are subjective to personal opinion.
Decide for yourself if this tool is useful (or not).
```

# GUIDE

This short guide will walk you through a practical example: E1M9 in `doom.wad`.

```
[doom.wad E1M9]
                  EASY  MEDIUM    HARD
hitscan %           24      36      33
health/AP           .5      .2      .1
armor/AP            .3      .1      .1
bullet/HP            4       2       1
shell/HP             2       2       2
rocket/HP           .6      .2      .2
plasma/HP            0       0       0
flags                p      p!      p!
```

**hitscan %**
This shows the percentage of monsters that use a hitscan attack. More hitscanners generally makes play harder, as these attacks cannot be dodged.

**health/AP**
This ratio is the total health pickups over total monster attack points. If this number is 1 it indicates the player can fully heal if every monster on the map gets one attack against the player, assuming median damage is dealt.

In E1M9 there is enough health items to heal 50% if all monsters attacked the player once for the easy skill, 20% on medium skill and 10% on hard.

**armor/AP**
This ratio is the total armor pickups over total monster attack points. If this number is 1 it indicates one armor point for every monster attack point on the map.

It should be noted that armor class (security/combat) is not factored in, and this ratio is strictly for comparative purposes to the baseline.

**bullet/HP, shell/HP, rocket/HP, plasma/HP**
Total damage for each munition type, over total monster hit points. A value of 1 indicates enough ammunition to kill each monster on the map at least once, assuming the median amount of damage is inflicted.

In E1M9 there are enough shells (factor of `2`) to kill each monster twice over.

**recommendation flags**
These are raised when the results fall outside the baseline. Use the `--legend` option to print the legend with the output.

```
AP: Monster attack points   HP: Monster hit points
 H: Health ratio too low     A: Armor ratio too low
 b: Bullet ratio too low     s: Shell ratio too low
 r: Rocket ratio too low     p: Plasma ratio too low
 !: Hitscan % too high
```

When tweaking a map to meet the baseline, use the `--compare` option to print `actual/baseline` values side-by-side:

```
$ dmon doom.wad E1M9 --compare
[doom.wad E1M9 versus DOOM2]
                  EASY  MEDIUM    HARD
hitscan %        24/38   36/35   33/32
health/AP        .5/.3   .2/.2   .1/.1
armor/AP         .3/.2   .1/.1   .1/.1
bullet/HP          4/1    2/.6    1/.4
shell/HP           2/2     2/1    2/.8
rocket/HP        .6/.5   .2/.2   .2/.2
plasma/HP          0/2    0/.8    0/.6
flags                p      p!      p!
```

Another way to distinguish variance is with the `--diff` option, which prints the difference between actual and baseline:

```
$ dmon doom.wad E1M9 --diff
[doom.wad E1M9 versus DOOM2]
                  EASY  MEDIUM    HARD
hitscan %          -14      +1     +.7
health/AP          +.2    -.03    +.02
armor/AP           +.1       0    -.03
bullet/HP           +3      +1     +.8
shell/HP           +.5      +1     +.9
rocket/HP          +.1       0       0
plasma/HP           -2      -1      -1
flags                p      p!      p!
```

## MORE EXAMPLES

Average for all DOOM2 maps:

    $ dmon DOOM2.WAD --average

Show stats for maps 1-4, counting bonus items, and print the legend:

    $ dmon DOOM2.WAD MAP0[1234] --legend --bonus

Dump all values to csv:

    $ dmon DOOM2.WAD --format=csv

# TESTS

A collection of unit tests are implemented for regression testing.

```
$ python tests.py -v
Baseline expected keys exist ... ok
Baseline values are numbers ... ok
Derive armor to monster ratio ... ok
Derive average ratios ... ok
Derive averages ... ok
Derive bullet damage to monster hit points ratio. ... ok
Derive health to monster ratio ... ok
Derive hitscanner percentage ... ok
Derive plasma damage to monster hit points ratio. ... ok
Derive rocket damage to monster hit points ratio. ... ok
Derive shell damage to monster hit points ratio. ... ok
Count armor points ... ok
Count armor points including bonus items ... ok
Count bullets ... ok
Extraction returns statistics object ... ok
Count health points ... ok
Count health points including bonus items ... ok
Count hitscanners ... ok
Count meaty monsters ... ok
Count monster hit points ... ok
Count plasma ... ok
Count rockets ... ok
Count shells ... ok
Count items with varied skill flags ... ok

----------------------------------------------------------------------
Ran 24 tests in 0.159s
```

# CREDITS

dmon uses [Omgifol](https://github.com/devinacker/omgifol), A Python library for manipulation of WAD files. Copyright (c) 2005 Fredrik Johansson, 2017 Devin Acker.

# LICENSE

Copyright (c) 2019 Wesley (beast) Werner

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
USE OR OTHER DEALINGS IN THE SOFTWARE.


changes.txt
Version 1.2 (November 2020)
- Ignore the --compare option if --diff option is set
- Pivot stats as columns instead of rows. Provides a new --pivot parameter for the old format.
- Update Omgifol + Python 3 compatibility
- Calculate bullet and shell ratio as damage over total monster hit points
- Derive armor and health bonuses against monster attack points
- Add rockets and plasma munition types
- Update baselines to use the new derivation calculations

Version 1.1 (August 2019)
- Initial release

src/lookup.py
"""
A lookup table for hitscanner types.
"""
hitscanner = {
    7: True, # spider mastermind
    9: True, # shotgun guy
    65: True, # commando/chaingunner
    3004: True, # zombieman
    84: True # ss guy
}


"""
A lookup table for non-hitscanner types.
"""
meaty = {
    3001: True, # imp
    3002: True, # demon
    58: True, # spectre
    3006: True, # lost soul
    3005: True, # cacodemon
    69: True, # hell knight
    3003: True, # baron of hell
    66: True, # revenant
    67: True, # mancubus
    68: True, # arachnotron
    71: True, # pain elemental
    64: True, # archvile
    16: True # cyberdemon
}


"""
Lookup table for health item by points.
"""
health = {
    2011: 10, # stimpack
    2012: 25, # medikit
    2014: 1 # health bonus
}


health_bonus = {
    2023: 100, # "berserk"
    2013: 100, # "supercharge"
    83: 200 # "megasphere"
}


"""
Lookup table for armor item by points.
"""
armor = {
    2015: 1, # armor bonus
    2018: 100, # green armor
    2019: 200 # blue armor
}


armor_bonus = {
    83: 200 # "megasphere"
}


"""
Count of shells per thing.
"""
shells = {
    2008: 4, # 4 shells
    2049: 20, # shell box
    2001: 8, # shotgun
    82: 8, # super shotgun
    9: 4, # shotgun guy - drops weapon on death
    8: 4 # backpack
}


"""
Count of bullets per thing.
"""
bullets = {
    2002: 20, # chaingun
    2007: 10, # ammo clip
    2048: 50, # ammo box
    3004: 5, # zombie - drops clip on death
    8: 10 # backpack
}


"""
Count of rockets per thing.
"""
rockets = {
    2003: 2, # rocket launcher
    2010: 1, # single rocket
    2046: 5 # box of rockets
}


"""
Count of plasma per thing.
"""
plasma = {
    2006: 40, # BFG
    17: 100, # cell pack
    2004: 40, # plasma rifle
    2047: 20 # energy cell
}


"""
Monster hit points lookup.
"""
monster_hp = {
    7: 3000, # spider mastermind
    9: 30, # shotgun guy
    16: 4000, # cyberdemon
    58: 150, # spectre
    64: 700, # archvile
    65: 70, # commando/chaingunner
    66: 30, # revenant
    67: 600, # mancubus
    68: 500, # arachnotron
    69: 500, # hell knight
    71: 400, # pain elemental
    84: 50, # ss guy
    3001: 60, # imp
    3002: 150, # demon
    3003: 1000, # baron of hell
    3004: 20, # zombieman
    3005: 400, # cacodemon
    3006: 100 # lost soul
}


"""
Monster attack damages lookup.
Damage is taken as the median damage per monster attack.
https://doomwiki.org/wiki/Spiderdemon#Data
"""
monster_ap = {
    7: 27, # spider mastermind
    9: 27, # shotgun guy
    16: 654, # cyberdemon
    58: 22, # spectre
    64: 55, # archvile
    65: 9, # commando/chaingunner
    66: 83, # revenant (30 punch + 50 missile)
    67: 32, # mancubus (per fireball)
    68: 25, # arachnotron
    69: 90, # hell knight (50 claw attack + 40 fireball)
    71: 75, # pain elemental (no attack, but assumes 5x lost sould spawns)
    84: 9, # ss guy
    3001: 12, # imp (melee and fireball attacks are the same)
    3002: 22, # demon
    3003: 90, # baron of hell (50 claw + 40 fireball)
    3004: 9, # zombieman
    3005: 30, # cacodemon
    3006: 15 # lost soul (goring attack)
}

src/omg/util.py
"""
    Utilities -- common functions and classes, used variously
    by other Omgifol modules.
"""

from fnmatch import fnmatchcase as wccmp
from struct  import pack, unpack, calcsize
from copy    import copy, deepcopy

_pack = pack
_unpack = unpack

class OrderedDict:
    """A dict-like container that remembers in which order items
    were added."""

    def __init__(self, source=None):
        """Create new, optionally from contents of given source."""
        self._items = {}
        self._n = 0
        if source:
            self.update(source)

    def __setitem__(self, key, value):
        """Set an item."""
        self._items[key] = self._n, value
        self._n += 1

    def __getitem__(self, key):
        """Retrieven an item."""
        return self._items[key][1]

    def __delitem__(self, key):
        """Delete an item."""
        del self._items[key]

    def __contains__(self, key):
        """Find if the dict holds the given key."""
        return key in self._items

    def __iter__(self):
        """Iterate over keys"""
        return iter(self.keys())

    def __add__(self, other):
        """Adds two dicts, copying items shallowly"""
        c = self.__class__()
        c.update(self)
        c.update(other)
        return c

    def __len__(self):
        """len(self)"""
        return len(self._items)

    def update(self, other):
        """Adds all content from another dictionary."""
        for k in other:
            self[k] = other[k]

    def items(self):
        """Returns a list of (key, value) tuples for all items."""
        data = [(v[0], k, v[1]) for (k, v) in self._items.items()]
        data.sort()
        return [(d[1], d[2]) for d in data]

    def keys(self):
        """Returns a list of all keys."""
        data = [(v[0], k, v[1]) for (k, v) in self._items.items()]
        data.sort()
        return [d[1] for d in data]

    def values(self):
        """Returns a list of all values."""
        data = [(v[0], k, v[1]) for (k, v) in self._items.items()]
        data.sort()
        return [d[2] for d in data]

    def clear(self):
        """Delete all items."""
        self._items.clear()

    def find(self, pattern):
        """Find all items that match the given pattern (supporting
        wildcards). Returns a list of keys."""
        return [k for k in self.keys() if wccmp(k, pattern)]

    def rename(self, old, new):
        """Rename an entry"""
        print old, new
        print old in self
        print "K", self.keys()
        self[new] = self[old]
        print self[new]
        del self[old]

    def __copy__(self):
        """Creates a deep copy."""
        a = self.__class__()
        for k in self:
            a[k] = copy(self[k])
        return a


#----------------------------------------------------------------------
#
# Miscellaneous convenient function
#

def join(seq):
    """Create a joined string out of a list of substrings."""
    return "".join(seq)

def readfile(source):
    """Read data from a file, return data as a string. Target may
    be a path name string or a file-like object (with a `read` method)."""
    if isinstance(source, str):
        return open(source, 'rb').read()
    else:
        return source.read()

def writefile(target, data):
    """Write data to a file. Target may be a path name string
    or a file-like object (with a `write` method)."""
    if isinstance(target, str):
        open(target,'wb').write(data)
    else:
        target.write(data)

def any(set):
    for e in set:
        if e:
            return True
    return False

def all(set):
    for e in set:
        if not e:
            return False
    return True

def inwclist(elem, seq):
    return any(wccmp(elem, x) for x in seq)


#----------------------------------------------------------------------
#
# Functions for processing lump names and other strings
#

# Table for translating characters to those safe for use
_trans_table = ["_"] * 256
for c in "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789[]\\_-":
    _trans_table[ord(c.lower())] = c
    _trans_table[ord(c)] = c
_trans_table[0] = "\0"
_trans_table = "".join(_trans_table)

def zpad(chars):
    """Pad a string with zero bytes, up until a length of 8.
    The string is truncated if longer than 8 bytes."""
    return pack('8s', chars)

def zstrip(chars):
    """Strip all data following the first zero in the string"""
    if '\0' in chars:
        return chars[:chars.index("\0")]
    return chars

def safe_name(chars):
    return chars[:8].translate(_trans_table)

def fixname(chars):
    if '\0' in chars:
        chars = chars[:chars.index("\0")]
    chars = chars.translate(_trans_table)
    return chars

def fixpadname(chars):
    """Same as fixname, but returns a string of exactly 8 bytes length,
    using zero (0x00) bytes for padding."""
    print "DEPRECATED!"
    return zpad(fixname(chars))

def fix_saving_name(name):
    """Neutralizes backslashes in Arch-Vile frame names"""
    return name.rstrip('\0').replace('\\', '`')

def fix_loading_name(name):
    """Restores backslash to Arch-Vile frame names"""
    return fixname(name).replace('`', '\\')

def unpack16(s):
    """Convert a packed signed short (2 bytes) to a Python int"""
    return unpack('h', s)[0]

def pack16(n):
    """Convert a Python int to a packed signed short (2 bytes)"""
    return pack('h', n)

def unpack32(s):
    """Convert a packed signed long (4 bytes) to a Python int"""
    return unpack('l', s)[0]

def pack32(n):
    """Convert a Python int to a packed signed long (4 bytes)"""
    return pack('l', n)


#----------------------------------------------------------------------
#
# A tool that can generate "Struct" classes for packing, unpacking, and
# representing the unpacked form of binary data.
#
# If this code looks ugly, be happy you didn't see the
# original version...
#

# Template for struct class generation
_struct_template = '''
class Struct(object):
    """%(doc)s"""

    _fmtsize = %(fmtsize)i
    _fmt  = %(fmt)r

    def __init__(self, %(initargs)s, bytes=None):
        if bytes:
            %(unpackexpr)s
        else:
            %(initbody)s
        %(init_exec)s

    def __getattribute__(self, name):
        return object.__getattribute__(self, name)

    def pack(self):
        return %(packexpr)s

%(flagdefs)s

Struct.__name__ = %(name)r
'''

# Template for struct flag property
_flagproperty = '''
    def get_%s(self):
        return bool(self.flags & %i)
    def set_%s(self, value):
        if value:
            self.flags |= %i
        else:
            self.flags &= %i
    %s = property(get_%s, set_%s)
'''

def _structdef(name, doc, fields, flags=None, init_exec=""):
    """Helper function for make_struct. Needed because Python doesn't
    like compile() and exec in the place when there are unknown
    variables floating around... (?)"""

    extra  = [f for f in fields if f[1] == 'x']
    fields = [f for f in fields if f[1] != 'x']

    # Wesley:
    # FIX: Force little-endian byte-order with the "<" token
    # to to work on *nix
    fmt = "<" + "".join(f[1] for f in fields)
    fmtsize = calcsize(fmt)

    # properties for easy access to the 'flags' bit field
    flagdefs = ""
    if flags:
        i = 1
        for f in flags:
            flagdefs += _flagproperty % (f, i, f, i, ~i, f, f, f)
            i <<= 1

    if init_exec: init_exec += ";"
    init_exec += '; '.join("self.%s=%s" % (f[0], f[0]) for f in extra)

    # example:  x=0, y=0, foo="BAR"
    initargs = ', '.join(f[0] + "=" + repr(f[2]) for f in fields+extra)

    # example:  self.x, self.y, self.foo = unpack('hh8s', bytes);
    #           self.foo = zstrip(safe_name(self.foo))
    unpackexpr =  ', '.join('self.'+f[0] for f in fields)
    unpackexpr += (" = unpack(%r, bytes); " % fmt)
    unpackexpr += "; ".join("self.%s=zstrip(safe_name(self.%s))" % \
        (f[0], f[0]) for f in fields if 's' in f[1])

    # example:  self.x=x; self.y=y; self.foo=foo
    initbody = "; ".join("self.%s=%s" % (f[0], f[0]) for f in fields)

    # example:  pack()
    packs = []
    for f in fields:
        if 's' in f[1]:
            packs.append("zpad(safe_name(self.%s))" % f[0])
        else:
            packs.append("self.%s" % f[0])
    packexpr = ("pack(%r, " % fmt) + ', '.join(packs) + ")"

    s = _struct_template % locals()

    # print s.replace("Struct", name)
    return compile(s, "<struct>", "exec")

def make_struct(*args, **kwargs):
    """Create a Struct class according to the given format"""
    exec _structdef(*args, **kwargs)
    return Struct


src/dmon.py
"""Dmon

Usage:
  dmon <wad> [options]
  dmon <wad> <pattern> [options]
  dmon (-h | --help | --about | --license)

Options:
  -h --help                 Show this screen.
  --version                 Show current version.
  --license                 Print the license.
  --about                   Show how this all works.
  <pattern>                 match map name (E1M2 or MAP02)
                            '?' and '*' are wildcards eg:
                                E1M* (all episode 1)
                                MAP0[135] (MAP01, MAP03 and MAP05)
  -a, --average             Combine results from multiple maps into
                            average amounts.
  -u, --bonus               Include berserk, soulsphere and megesphere
                            as data points.
  -f <fmt>, --format=<fmt>  Set output format to: csv, json, dump
  -b <bl>, --baseline=<bl>  Set comparison baseline [Default: DOOM2]
                            where <bl> can be:
                                DOOM, DOOM2, SIGIL,
                                AV (Alien Vendetta),
                                HR (HELL REVEALED),
                                DTWID (Doom the Way id Did)
  -c, --compare             Show actual/baseline values for comparison.
  -d, --diff                Display the difference between actual and
                            baseline values.
  -x, --fixed               Display fixed-point numbers (otherwise
                            numbers are rounded up).
  -l, --legend              Print the recommendation legend at the end.
  -p, --pivot               Pivot the output so that statistics are listed
                            as columns instead of rows.
                            Only applies when --format is not set.
  --verbose                 Output debug messages.

"""


import sys
import os.path
from docopt import docopt
import constants
import baselines
import dmoncommon

__version__ = constants.VERSION


def test_file_exists(options):
    filename = options["<wad>"]
    exists = os.path.isfile(filename)
    if not exists:
        print("WAD not found: %s" % (filename))
    return exists


def process_wad(options):

    if not test_file_exists(options):
        return

    wad_data = dmoncommon.extract_statistics(options)

    fmt = options["--format"]
    if fmt is not None:
        fmt = fmt.upper()
    
    if len(wad_data["map list"]) == 0:
        map_pattern = options["<pattern>"] or "*"
        print("No maps in WAD, or none matched pattern '%s'."
        " An Imp tears your arm off." % (map_pattern))
    elif fmt is None:
        if options["--average"] == True:
            print(to_tabular(wad_data, options, True))
        else:
            print(to_tabular(wad_data, options, False))
    elif fmt == "CSV":
        print(to_csv(wad_data, options))
    elif fmt == "JSON":
        print(to_json(wad_data))
    elif fmt == "DUMP":
        print(to_baseline_dump(wad_data))
    else:
        print("Invalid output format '%s'."
        " A Demon bites your face off." % (fmt))

    if options["--legend"] == True:
        print_legend_flags(options)


def format_as_table(columns, rows, first_column_width=6, other_column_width=12):
    """
    Formats fixed-width columns.
    """
    fmts = ""
    for i, c in enumerate(columns):
        if i == 0:
            # first column is left-aligned
            fmts += "{:<"+str(first_column_width)+"}"
        else:
            # others right-aligned
            fmts += "{:>"+str(other_column_width)+"}"
    output = fmts.format(*columns) + "\n"
    for values in rows:
        output += fmts.format(*values) + "\n"
    return output


def to_tabular(wad_data, options, list_averages):
    """
    Generates a tabular output string from generated statistics.
    Each map is generated in turn (where PATTERN is matched).
    If list_averages is Truthy then the average data is used and
    the map loop is broken early.
    """

    # column titles for each skill
    column_titles = ["", "EASY", "MEDIUM", "HARD"]
    
    # Format width of first column.
    # This will be smaller for pivoted output.
    first_column_width = 14
    other_column_width = 8

    cmp_mode = options["--compare"]
    diff_mode = options["--diff"]
    pivoted_mode = options["--pivot"]
    output = ""
    
    # Increase other columns for compare mode
    if cmp_mode:
        other_column_width = 12
    
    # List of maps to print
    map_list = wad_data["map list"]
    
    # Overwrite the map list when printing averages
    if list_averages:
        map_list = ["AVERAGES",]

    for map_name in map_list:
        map_data = wad_data["results"][map_name]
        
        # Print file and map names
        if cmp_mode or diff_mode:
            fmt_values = (wad_data["filename"], map_name, options["--baseline"])
            output += "\n[%s %s versus %s]\n" % fmt_values
        else:
            fmt_values = (wad_data["filename"], map_name)
            output += "\n[%s %s]\n" % fmt_values

        rows = []
        
        if pivoted_mode:
            # List statistics as columns
            column_titles = ("skill",) + constants.TITLES
            # Adjust column sizes for pivoted output
            first_column_width = 6
            other_column_width = 14
            for skill in constants.SKILLS:
                # Add the skill title
                this_row = [skill,]
                for i, title in enumerate(constants.TITLES):
                    # Average row has no flags, default to blank
                    stat_value = map_data[skill].get(title) or ""
                    this_row.append(stat_value)
                rows.append(this_row)
        else:
            # List statistic as rows
            for i, title in enumerate(constants.TITLES):
                # Add the stat title
                this_row = [title,]
                # Add the values for this stat from each skill level
                for skill in constants.SKILLS:
                    # Average row has no flags, default to blank
                    stat_value = map_data[skill].get(title) or ""
                    this_row.append(stat_value)
                rows.append(this_row)

        output += format_as_table(column_titles, rows,
                    first_column_width, other_column_width)

    return output


def to_csv(wad_data, options):
    """
    Output statistics as CSV data.
    """

    csv = (
        "FILE,MAP,"
        "SKILL,"
        "MONSTERS,"
        "HITSCANNERS,"
        "HEALTH POINTS,"
        "ARMOR POINTS,"
        "BULLETS,"
        "SHELLS,"
        "ROCKETS,"
        "CELLS,"
    )
    
    # Add derived stat columns
    for col in constants.TITLES:
        if col != "flags":
            csv += col.upper() + ","

    # Trim trailing comma and end the line
    csv = csv[:-1] + "\n"

    # Add each map and skill as a row
    for map_name in wad_data["map list"]:
        map_data = wad_data["data"][map_name]
        for skill in constants.SKILLS:
            skill_data = map_data[skill]
            csv += wad_data["filename"] + ","
            csv += map_name + ","
            csv += skill + ","
            csv += str(skill_data["monsters"]) + ","
            csv += str(skill_data["hitscanners"]) + ","
            csv += str(skill_data["health points"]) + ","
            csv += str(skill_data["armor points"]) + ","
            csv += str(skill_data["bullets"]) + ","
            csv += str(skill_data["shells"]) + ","
            csv += str(skill_data["rockets"]) + ","
            csv += str(skill_data["plasma cells"]) + ","
            # Add each derived value
            for col in constants.TITLES:
                if col != "flags":
                    csv += str(skill_data[col]) + ","
            # Trim trailing comma and end the line
            csv = csv[:-1] + "\n"

    return csv


def to_json(wad_data):
    """
    Output statistics as raw data.
    """
    import json
    return json.dumps(wad_data, indent=2)


def to_baseline_dump(wad_data):
    """
    Output as baseline-structured dump.
    """
    average_data = wad_data["data"]["AVERAGES"]
    for skill in constants.SKILLS:
        for key, value in average_data[skill].items():
            average_data[skill][key] = round(value, 1)

    import json
    return json.dumps(average_data, indent=4)


def print_legend_flags(options):
    """
    Print the list of recommendation flags with baseline values.
    """
    print("AP: Monster attack points   HP: Monster hit points")
    print(" %s: Health ratio too low     %s: Armor ratio too low"
        % (constants.RECOMMENDS_HP, constants.RECOMMENDS_AP))
    print(" %s: Bullet ratio too low     %s: Shell ratio too low"
        % (constants.RECOMMENDS_BULLETS, constants.RECOMMENDS_SHELLS))
    print(" %s: Rocket ratio too low     %s: Plasma ratio too low"
        % (constants.RECOMMENDS_ROCKETS, constants.RECOMMENDS_PLASMA))
    print(" %s: Hitscan %% too high" % (constants.RECOMMENDS_HITSCANS))


def process_help_and_about_options(options):

    if options["--license"] == True:
        print(constants.LICENSE)
        return True

    if options["--about"] == True:
        print(constants.ABOUT)
        return True

    return False


def valid_baseline_specified(options):
    code = options["--baseline"].upper()
    if code in baselines.lookup:
        return True
    else:
        print("%s is not a valid baseline. "
        "A Cyberdemon blasts you to pieces." % (code))
        return False


def main():
    options = docopt(__doc__, version=__version__)
    if isinstance(options, dict):
        if not valid_baseline_specified(options):
            sys.exit(255)

        if not process_help_and_about_options(options):
            process_wad(options)
    else:
        print(options)


if __name__ == "__main__":
    main()
